src/common/src/types/interval.rs

```diff
@@ -485,20 +516,154 @@ impl std::fmt::Debug for IntervalUnitDisplay<'_> {
+/// deserialize from `IntervalCmpValue`.
+#[derive(PartialEq, Eq, Hash, PartialOrd, Ord)]
+struct IntervalCmpValue(i128);
+
+impl From<IntervalUnit> for IntervalCmpValue {
+    fn from(value: IntervalUnit) -> Self {
+        let days = (value.days as i64) + 30i64 * (value.months as i64);
+        let usecs = (value.usecs as i128) + (USECS_PER_DAY as i128) * (days as i128);
+        Self(usecs)
+    }
+}
+
+impl Ord for IntervalUnit {
+    fn cmp(&self, other: &Self) -> Ordering {
+        IntervalCmpValue::from(*self).cmp(&(*other).into())
+    }
+}
+
+impl PartialOrd for IntervalUnit {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for IntervalUnit {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other).is_eq()
+    }
+}
+
+impl Eq for IntervalUnit {}
+
+impl Hash for IntervalUnit {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        IntervalCmpValue::from(*self).hash(state);
+    }
+}
+
+/// Loss of information during the process due to `IntervalCmpValue`. Only intended for
+/// memcomparable encoding.
 impl Serialize for IntervalUnit {
     fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
     where
         S: serde::Serializer,
     {
-        let IntervalUnit {
-            months,
-            days,
-            usecs,
-        } = self.justified();
-        // serialize the `IntervalUnit` as a tuple
-        (months, days, usecs).serialize(serializer)
+        let cmp_value = IntervalCmpValue::from(*self);
+        // split i128 as (i64, u64), which is equivalent
+        (
+            (cmp_value.0 >> 64) as i64,
+            cmp_value.0 as u64, // truncate to get the lower part
+        )
+            .serialize(serializer)
+    }
+}
+
+impl IntervalCmpValue {
+    /// Recover the justified interval from this equivalence class, if it exists.
+    fn as_justified(&self) -> Option<IntervalUnit> {
+        let usecs = (self.0 % (USECS_PER_DAY as i128)) as i64;
+        let remaining_days = self.0 / (USECS_PER_DAY as i128);
+        let days = (remaining_days % 30) as i32;
+        let months = (remaining_days / 30).try_into().ok()?;
+        Some(IntervalUnit::from_month_day_usec(months, days, usecs))
+    }
+
+    /// Recover the alternate representative interval from this equivalence class.
+    /// It always exists unless the encoding is invalid. See [`IntervalCmpValue`] for details.
+    fn as_alternate(&self) -> Option<IntervalUnit> {
+        match self.0.cmp(&0) {
+            Ordering::Equal => Some(IntervalUnit::from_month_day_usec(0, 0, 0)),
+            Ordering::Greater => {
+                let remaining_usecs = self.0;
+                let mut usecs = (remaining_usecs % (USECS_PER_DAY as i128)) as i64;
+                let mut remaining_days = remaining_usecs / (USECS_PER_DAY as i128);
+                let extra_days = ((i64::MAX - usecs) / USECS_PER_DAY)
```

Comment:
I see, this handles overflow?

Reply:
Not sure what the question is exactly, but I will add more debug_asserts and comments to make it clear.